apps/rush-lib/src/cli/actions/ChangeAction.ts

```diff
@@ -25,9 +25,10 @@ import {
   LockStepVersionPolicy,
   VersionPolicyDefinitionName
 } from '../../api/VersionPolicy';
+import { ProjectChangeAnalyzer } from '../../logic/ProjectChangeAnalyzer';
```

Comment:
When writing the docs for [Build cache](https://rushjs.io/pages/maintainer/build_cache/) and [Incremental builds](https://rushjs.io/pages/advanced/incremental_builds/), I struggled with the right words for several confusingly similar concepts:

1. The engine that determines whether a project can be retrieved from the build cache
2. The engine that determines whether source code has been modified since the last time a folder was built
3. The `rush change` engine that determines which projects a PR branch has touched, as compared to `master`

These same topics are closely related to deployment/test optimizations, where people as:
- Can we only run automation tests for projects that got rebuilt by Rush?
- Can we deploy only those projects that were modified by this PR branch?

I referred to 1 and 2 as the "incremental analyzer" and pretended they are one entity.  (Because we anticipate that the build cache might eventually entirely replace the **package-deps-hash** optimization.)

`ProjectChangeAnalyzer` sounds more like 3.  Would we have a separate API for 1 and 2?  Or would it be natural to roll all these things into one centralized API (in which case maybe it should be called `IncrementalAnalyzer` instead of `ProjectChangeAnalyzer`)?

CC @elliot-nelson

Reply:
I might be misunderstanding your note @octogonz, but I don't think you had to pretend -- both the build cache and the task selector / build watcher all use the PackageChangeAnalyzer and in the same way.  So 1 and 2 are the same entity, just applied at different times and with different results.

3 (rush change) is different, that uses the git client directly today to ask questions about files that are changed and doesn't grab the list of hashes.  There might be reasons to make 3 the same as 1 and 2 for simplification, but it's possible that would be a performance hit for rush change (I don't know the relative performance of the getChangedFiles call vs package-deps-hash module off hand).

Reply:
If I understand right:

1. Compares the hash-of-project-hashes string against the build cache key string.  It is asking ***"do we have a cache entry for these inputs?"***  Oddly it overwrites the output files even if they are up to date.
2. Compares the project hashes against `.rush/temp/package-deps_build.json`. I suspect that this logic is only used when the build cache experiment is disabled.  It is asking ***"has the code changed since when we last compiled this project?"***  Oddly it does not even look at the output files to see if they are there.
3. Is asking ***"does the Git branch include changes to this project?"*** That seems like a separate problem that is not solvable by 1+2.

@iclanton is this right?

Reply:
Ah ok I see what you mean now.

I think 2 is used even if build cache is enabled though -- as an example, make a local change to a project and build, then build again.  The second time will be a cache miss (for 1) but a hash match (for 2), resulting in a "skipped" project instead of a "retrieved from build cache" project.

Reply:
1 and 2 largely use the same mechanism for determining a project's state. The big difference is that 1 generates a hash of the hashes and 2 does a textual diff of the serialized map (with some other caveats for things like CLI arguments).

3 does indeed use a different mechanism for determining if the project has changed. Namely, it gets a list of all paths that have changed relative to the main/master branch and then checks which projects they are in.